ha/setup/ha_setup.py

```diff
@@ -0,0 +1,343 @@
+from ha import const
+#from ha.setup.create_cluster import cluster_auth, cluster_create
+#from ha.setup.create_pacemaker_resources import create_all_resources
+
+class Cmd:
+    """
+    Setup Command. This class provides methods for parsing arguments.
+    """
+    _index = "conf"
+
+    def __init__(self, args: dict):
+        """
+        Init method.
+        """
+        self._url = args.config
+        Conf.load(self._index, self._url)
+        self._args = args.args
+        self._execute = SimpleCommand()
+
+    @property
+    def args(self) -> str:
+        return self._args
+
+    @property
+    def url(self) -> str:
+        return self._url
+
+    @staticmethod
+    def usage(prog: str):
+        """
+        Print usage instructions
+        """
+        sys.stderr.write(
+            f"usage: {prog} [-h] <cmd> <--config url> [--plan <TYPE>] <args>...\n"
+            f"where:\n"
+            f"cmd   post_install, config, init, test, upgrade, reset, cleanup, backup, restore\n"
+            f"--config   Config URL\n"
+            f"--plan   Test plan")
+
+    @staticmethod
+    def get_command(desc: str, argv: dict):
+        """
+        Return the Command after parsing the command line.
+        """
+        parser = argparse.ArgumentParser(desc)
+
+        subparsers = parser.add_subparsers()
+        cmds = inspect.getmembers(sys.modules[__name__])
+        cmds = [(x, y) for x, y in cmds
+            if x.endswith("Cmd") and x != "Cmd"]
+        for name, cmd in cmds:
+            cmd.add_args(subparsers, cmd, name)
+        args = parser.parse_args(argv)
+        return args.command(args)
+
+    @staticmethod
+    def add_args(parser: str, cls: str, name: str):
+        """
+        Add Command args for parsing.
+        """
+        parser1 = parser.add_parser(cls.name, help='setup %s' % name)
+        parser1.add_argument('--config', help='Config URL')
+        parser1.add_argument('--plan', nargs="?", default=[], help='Test plan')
+        parser1.add_argument('args', nargs='*', default=[], help='args')
+        parser1.set_defaults(command=cls)
+
+class PostInstallCmd(Cmd):
+    """
+    PostInstall Setup Cmd
+    """
+    name = "post_install"
+
+    def __init__(self, args: dict):
+        """
+        Init method.
+        """
+        super().__init__(args)
+
+    def process(self):
+        """
+        Process post_install command.
+        """
+
+        Log.info("Processing post_install command")
+        # Create a directory and copy config file
+        if os.path.exists(const.HA_CONFIG_FILE):
+            os.remove(const.HA_CONFIG_FILE)
+        os.makedirs(const.CONFIG_DIR, exist_ok=True)
+        shutil.copyfile(const.SOURCE_CONFIG_FILE, const.HA_CONFIG_FILE)
+        Log.info(f"{self.name}: Copied HA config file.")
+        # Pre-requisite checks are done here.
+        # Make sure the pacemaker, corosync and pcs packages have been installed
+        PkgV().validate('rpms', const.PCS_CLUSTER_PACKAGES)
```

Comment:
I guess, exception should be caught, log the error and them raise HaPrerequisiteException?

Reply:
The exception is caught in main below which logs the error and also displays which command had failed. And it is applicable to all stages in all of these commands.

Reply:
Added HaPrerequisiteException exception as well.